src/cargo/sources/registry/mod.rs

```diff
@@ -739,7 +739,8 @@ impl<'cfg> Source for RegistrySource<'cfg> {
             .query_inner(dep, &mut *self.ops, &self.yanked_whitelist, &mut |s| {
                 let matched = match kind {
                     QueryKind::Exact => dep.matches(&s),
-                    QueryKind::Fuzzy => true,
+                    QueryKind::Alternatives => true,
+                    QueryKind::Normalized => true,
```

Comment:
> The test now replicates the original error and makes an assertion of what the correct output should be. It's still not passing though.

Put an assert in here and run your tests.  My guess is this code isn't being hit.

My guess is we are "publishing" the packages to a directory registry which will be hitting the `directory.rs` code path rather than the `registry/mod.rs` code path.

**maybe** if we setup a sparse/http registry in the test, we could reproduce this.

Another alternative is we go past the MVP solution and implement normalization for all sources.

The final alternative is that we don't have a test for this.  As the note above hints at, we are only testing one source type's normalization and there would be two main classes of normalization here.  Unless we went and verified all sources, there isn't really much of a difference between implementing the feature for a different code path and testing it vs just not testing this.

Reply:
note: also as part of characterizing this, have you been able to reproduce the expected behavior by hand by running `cargo run -- add ...`?

Reply:
So both the test and running by hand via `../<snip>/target/debug/cargo add ...` are hitting the code in `src/cargo/sources/registry/mod.rs` actually. I am not able to reproduce the expected behavior by hand, I still get the original error. So seems like I am missing something in the implementation here. I need to actually dig into it more and get some context which might take me a day or two more. I'll try to come back with a solution or a pointed question.

Reply:
Update: Haven't given up at all, but had to pack up my family members house being sold this week. Will have time to dig in this weekend.